src/dmd/expressionsem.d

```diff
@@ -8826,12 +8826,41 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
 
                 // May be block or element-wise assignment, so
                 // convert e1 to e1[]
-                // flatten into a 1D array unless e2 is an array type
-                if (exp.op != TOK.assign && e2x.type.ty != Tarray && e2x.type.ty != Tsarray)
-                {
-                    // If multidimensional static array, treat as one large array
-                    const dim = t1.numberOfElems(exp.loc);
-                    e1x.type = t1.baseElemOf().sarrayOf(dim);
+                // If multidimensional static array, treat as one large array
+                //
+                // Find the appropriate array type depending on the assignment, e.g.
+                // int[3] = int => int[3]
+                // int[3][2] = int => int[6]
+                // int[3][2] = int[] => int[3][2]
+                // int[3][2][4] + int => int[24]
+                // int[3][2][4] + int[] => int[3][8]
+                ulong dim = 1;
+                auto type = e1x.type;
```

Comment:
Since the first level is given to be a static array, you can take off one cycle from the loop.
```suggestion
                ulong dim = t1.isTypeSArray().dim.toUInteger();
                auto type = t1.nextOf();
```

Reply:
Done